src/dust.ts

```diff
@@ -0,0 +1,196 @@
+        }
+        this._Model = Model;
+        this._particles = [];
+        this._random = random;
+
+        // NOTE: Here, we construct the model just once per dust
+        // object and then use exactly the same model object
+        // (including temporary internal storage) across all particles
+        // in object. This is in contrast with the approach taken in
+        // the C++ code where we kept separate copies so that they
+        // could be run in parallel. If we parallelise this with web
+        // workers we might need to revisit this approach, but the API
+        // there looks nothing like OpenMP and this would probably
+        // still be ok, once per worker.
+        const model = new Model(base, pars);
+        for (let i = 0; i < nParticles; ++i) {
+            this._particles.push(new Particle(model, step));
+        }
+    }
+
+    /**
+     * Returns the number of state elements per particle
+     */
+    public nState(): number {
+        return this._particles[0].size;
+    }
+
+    /**
+     * Returns the number of particles
+     */
+    public nParticles(): number {
+        return this._particles.length;
+    }
+
+    /**
+     * The current step
+     */
+    public step(): number {
+        return this._particles[0].step;
+    }
+
+    /**
+     * Returns information about how the state is packed. This is
+     * useful for computing indexes and extracting state.
+     */
+    public info(): DustModelInfo {
+        return this._particles[0].info();
+    }
+
+    /**
+     * Set new parameters into the model
+     *
+     * @param pars: New parameters
+     *
+     * @param setInitialState If `true`, then changing the parameters
+     * also updates the initial conditions to use those from the
+     * model's `initial()` method. This uses the current step of the
+     * model, so you may want to use {@link Dust.setStep} first.
+     */
+    public setPars(pars: Pars, setInitialState: boolean): void {
+        const step = this.step();
+        const nState = this.nState();
+        const model = new this._Model(base, pars);
+        if (model.size() !== this.nState()) {
+            throw Error(`Particle produced unexpected state size`);
+        }
+        for (let i = 0; i < this._particles.length; ++i) {
+            const state =
+                setInitialState ? undefined : this._particles[i].state();
+            this._particles[i] = new Particle(model, step, state);
+        }
+    }
+
+    /**
+     * Change the step in the model, without updating state
+     *
+     * @param step New step
+     */
+    public setStep(step: number): void {
+        this._particles.forEach((p) => {
+            p.step = step;
+        });
+    }
+
+    /**
+     * Change the model state
+     *
+     * @param state A 2d-matrix of state; this inteface may change.
+     */
+    public setState(state: number[][]): void {
+        this.checkState(state);
+        for (let i = 0; i < this.nParticles(); ++i) {
```

Comment:
I wonder if it would be worth making a helper method for iterating through particles, since that seems to happen quite a bit. Wouldn't really save many lines but might look a little cleaner. something like:
```
private forEachParticle(do: (p: Particle, idx: number) => void) => {
   this._particles.forEach(do);
});
```
then e.g. `this.forEachParticle((p: Particle, idx: number) => { p.setState(state[idx]); })`

Reply:
Done now in a2a6073, I had no idea `for` loops were so disliked :)